### PR TITLE
Fips 203 final - updates to code, proof, and KATs

### PR DESCRIPTION
**Pull request by rod-chapman**
This PR updates this implementation for FIPS 203 FINAL issued 13th August 2024.

Proofs with GNATProve are 100% OK.

Updates to KATs will be cross-checked with other implementations ASAP.

### CODE REVIEW

**Comment by rod-chapman** [comment]
Code changes are mostly harmless. The new input validation functions (FIPS 203 Section 7) should be reviewed carefully.

File: spark_ada/src/mlkem.adb [hunk 1]
```diff
@@ -1945,6 +1947,96 @@ is
...
+   end Key_Pair_Is_Consistent_With_Seed;
+
+
+   --  FIPS 203 section 7.1 "Key pair check (without seed)"
+   function Key_Pair_Check_Without_Seed
+      (Key_Pair : in MLKEM_Key;
+       Random_M : in Bytes_32) return Boolean
+   is
+      Key          : Bytes_32;
+      Key_Tick     : Bytes_32;
```
Comment (dkostic) [inline spark_ada/src/mlkem.adb:2009]: nit: `Key` is confusing to me here, I prefer `Shared_Secret` or something like that.

File: spark_ada/src/mlkem.adb [hunk 2]
```diff
@@ -1945,6 +1947,96 @@ is
...
+      Test, Reference_Hash : Bytes_32;
+   begin
+      --  FIPS 203 7.3 - Decapsulation key check
+      --    1. Ciphertext type check. This is a static type-check in SPARK,
+      --       so nothing to do here.
+      --    2. Decapsulation key type check. This is a static type-check in
+      --       SPARK, so nothing to do here.
+      --    3. Hash check, as follows:
+      Test           := H (HD);
+      Reference_Hash := Bytes_32 (DK_Bar (768 * K + 32 .. 768 * K + 63));
```
Comment (dkostic) [inline spark_ada/src/mlkem.adb:1984]: same comment as above.

File: spark_ada/src/mlkem.adb [hunk 3]
```diff
@@ -1964,35 +2056,20 @@ is
...
-      Reencoded := ByteEncode12 (Decoded);
-      return Byte_Seq_Equal (Key_To_Check, Reencoded);
-   end EK_Is_Valid_For_Encaps;
-
-
+   -- This is also ML-KEM.Encaps_internal from FIPS 203 Algorithm 17
    procedure MLKEM_Encaps (EK       : in     MLKEM_Encapsulation_Key;
                            Random_M : in     Bytes_32;
-                           SS       :    out Bytes_32;
+                           Key      :    out Bytes_32;
```
Comment (dkostic) [inline spark_ada/src/mlkem.adb:2062]: funny, I see you prefer `Key` to `SS`

File: spark_ada/src/mlkem.adb [hunk 4]
```diff
@@ -1945,6 +1947,96 @@ is
    --  Exported subprogram bodies
    --=======================================
 
+   --  FIPS 203 section 7.2
+   function EK_Valid_For_Encaps (EK_Bar : in MLKEM_Encapsulation_Key)
+     return Boolean
+   is
+      Key_To_Check : constant Poly_Zq_Vector_Bytes := EK_Bar (0 .. 384 * K - 1);
```
[the remark this answers is not in the document]
Comment (dkostic) [reply spark_ada/src/mlkem.adb:1954]: ah, I just checked Mathematical Symbols part of the spec and it states:
```
𝐵[𝑘 ∶ 𝑚] The subarray (𝐵[𝑘], 𝐵[𝑘 + 1], ... , 𝐵[𝑚 − 1]) of the array 𝐵.
```
so I guess it's just abuse of notation.

**Review by dkostic — APPROVED** [review]
lgtm, just couple of nits.

File: spark_ada/src/mlkem.adb (hunk 4 above)
[the remark this answers is not in the document]
Comment (rod-chapman) [reply spark_ada/src/mlkem.adb:1954]: Yeah... notational difference... in Ada, a "slice" of an array is inclusive at both ends.

File: spark_ada/src/mlkem.adb (hunk 1 above)
Comment (rod-chapman) [reply spark_ada/src/mlkem.adb:2009]: The spec (section 7.2, para 4ii - 4iii) calls these values K and K', but "K" is already used to represent the algorithm parameter K, and ' isn't legal in identifiers, so I went for Key and Key_Tick.

File: spark_ada/src/mlkem.adb (hunk 3 above)
Comment (rod-chapman) [reply spark_ada/src/mlkem.adb:2062]: Again, in the spec they use "K" (Algorithm 20), which is overloaded, so I went for "Key"

**Comment by rod-chapman** [comment]
Thanks

File: spark_ada/src/mlkem.adb [hunk 5]
```diff
@@ -1945,6 +1947,96 @@ is
...
+   end EK_Valid_For_Encaps;
+
+   --  FIPS 203 section 7.3 - Decapsulation key and Ciphertext check
+   function DK_Valid_For_Decaps (DK_Bar : in MLKEM_Decapsulation_Key)
+     return Boolean
+   is
+      subtype Hash_Data_Index is I32 range 0 .. 384 * K + 31;
+      subtype Hash_Data is Byte_Seq (Hash_Data_Index);
+
+      HD : constant Hash_Data := Hash_Data (DK_Bar (384 * K .. 768 * K + 31));
```
[the remark this answers is not in the document]
Comment (rod-chapman) [reply spark_ada/src/mlkem.adb:1974]: As above, slices are inclusive in Ada.

**Pull request merged**